apps/jsonrpc2/test/jsonrpc2/handler_test.exs

```diff
@@ -0,0 +1,240 @@
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s([1]),
+        ~s([
+          {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null}
+        ])
+      )
+    end
+
+    test "rpc call with invalid Batch" do
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s([1,2,3]),
+        ~s([
+          {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null},
+          {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null},
+          {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null}
+        ])
+      )
+    end
+
+    test "rpc call Batch" do
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s([
+            {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
+            {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},
+            {"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},
+            {"foo": "boo"},
+            {"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"},
+            {"jsonrpc": "2.0", "method": "get_data", "id": "9"}
+        ]),
+        ~s([
+            {"jsonrpc": "2.0", "result": 7, "id": "1"},
+            {"jsonrpc": "2.0", "result": 19, "id": "2"},
+            {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null},
+            {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found", "data": {"method": "foo.get", "params": {"name": "myself"}}}, "id": "5"},
+            {"jsonrpc": "2.0", "result": ["hello", 5], "id": "9"}
+        ])
+      )
+    end
+
+    test "rpc call Batch (all notifications)" do
+      assert_rpc_noreply(
+        SpecHandlerTest,
+        ~s([
+          {"jsonrpc": "2.0", "method": "notify_sum", "params": [1,2,4]},
+          {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]}
+        ])
+      )
+    end
+  end
+
+  describe "internal tests" do
+    test "rpc call exit/raise/throw produces internal error" do
+      capture =
+        capture_log(fn ->
+          assert_rpc_reply(
+            JSONRPC2.ErrorHandlerTest,
+            ~s([
+                {"jsonrpc": "2.0", "method": "exit", "id": "1"},
+                {"jsonrpc": "2.0", "method": "raise", "id": "2"},
+                {"jsonrpc": "2.0", "method": "throw", "id": "3"}
+            ]),
+            ~s([
+                {"jsonrpc": "2.0", "id": "1", "error": {"message": "Internal error", "code": -32603}},
+                {"jsonrpc": "2.0", "id": "2", "error": {"message": "Internal error", "code": -32603}},
+                {"jsonrpc": "2.0", "id": "3", "error": {"message": "Internal error", "code": -32603}}
+              ])
+          )
+        end)
+
+      assert capture =~
+               "[error] Error in handler JSONRPC2.ErrorHandlerTest for method exit with params: []:"
```

Comment:
Yes, the test would fail if you change the format and I think it's desired.

Reply:
Yeah, it might be a bit too in the weeds. The purist in me thinks that if, in the text of the error message, `params` is changed to `parameters`, the functionality still works, so the test should still pass, but it will fail. In the unlikely event that this happens, we can worry about it then.

Reply:
hm... yeah, I know what you mean. I always stick to the convention that says: unit tests should check the output if X is the input. Strict input-output validation.
Acceptance tests on the other hand - I would use your approach. Is the server returning 503 - yes/no thats OK, I'm not interested in the message.